### PR TITLE
Pixflat

### PR DESCRIPTION
**Pull request by julienguy**
Pixel flat field script , using the new flat field slit data. This code has been used for the flats of SM1-SM8
saved as ```pixflat*.fits``` in https://desi.lbl.gov/trac/browser/calib/desi_spectro_calib/trunk/ccd

```
desi_compute_pixel_flatfield --help
usage: desi_compute_pixel_flatfield [-h] -i [IMAGES [IMAGES ...]] -o OUTFILE
                                    [--out-mean OUT_MEAN] [-d]
                                    [--minflux MINFLUX] [--maxerr MAXERR]

Computes a pixel level flat field image from a set of preprocessed images
obtained with the flatfield slit. An average image is computed if several
preprocessed images are given. The method consists in iteratively dividing the
image by a smoothed version of the same image, flat(n+1) =
flat(n)/smoothing(flat(n)). The smoothing consists in 1D median filtering
along the wavelength dispersion axis or the fiber axis alternatively, with
masking.

optional arguments:
  -h, --help            show this help message and exit
  -i [IMAGES [IMAGES ...]], --images [IMAGES [IMAGES ...]]
                        path to input preprocessed image fits files, or a
                        single median image (default: None)
  -o OUTFILE, --outfile OUTFILE
                        output flatfield image filename (default: None)
  --out-mean OUT_MEAN   save median image (default: None)
  -d, --debug           write a lot of debugging images (default: False)
  --minflux MINFLUX     minimum flux fraction of median (default: 0.05)
  --maxerr MAXERR       maximum error (default: 0.02)
```

### CODE REVIEW

**Comment by julienguy** [comment]
I am working on documenting the code.

You can find all the pixel flats in https://desi.lbl.gov/trac/browser/calib/desi_spectro_calib/trunk/ccd , files pixflat*.fits.

**Comment by profxj** [comment]
Here is a screenshot of a portion of sm5, z-band.
Do we understand the waves?  Just checking.

sm5-z

**Comment by julienguy** [comment]
This is likely fringing. It is not generated by the code. There are also larger quasi-periodic patterns that are likely tree rings (due to variations of dopant density). There are also glue bubbles visible in some flats in R and Z.

**Comment by julienguy** [comment]
@profxj , please have a second look at the code; and thanks for checking this.

File: bin/desi_compute_pixel_flatfield [hunk 1]
```diff
@@ -1,95 +1,55 @@
...
-    return x,y,dxdy,dydx
-
-def median_image(image_filenames) :
-    """ Return a median of input images after rescaling each image
+    return ndi.generic_filter(image, LowLevelCallable(maskedmed.ctypes), size=shape)
+
+
+
+
+def clipped_mean_image(image_filenames) :
```
Comment (profxj) [inline bin/desi_compute_pixel_flatfield]: I find it odd we are re-creating a basic task in `astropy`:

https://docs.astropy.org/en/stable/api/astropy.stats.sigma_clip.html

Odds are theirs is better tested and debugged..

File: bin/desi_compute_pixel_flatfield [hunk 2]
```diff
@@ -231,397 +219,99 @@ def gaussian_smoothing_1d_per_axis(image,ivar,sigma,npass=2,dxdy=0.,dydx=0.) :
...
+        flat  += (model<=minflat)|(ivar<=0)
+        log.info("step 4")
+        tmp = scipy.signal.medfilt2d(flat,shape_2)
+        model[gradmask] *= tmp[gradmask]
+        flat  =  (ivar>0)*(model>minflat)*image/(model*(model>minflat)+(model<=minflat))
+        flat  += (model<=minflat)|(ivar<=0)
+    return flat,model
+
+def amplifier_matching(image) :
+
```
Comment (profxj) [inline bin/desi_compute_pixel_flatfield]: can this method be briefly documented?

**Comment by julienguy** [comment]
Merging now because there is a potential issue with neg. pixel flat solved in this branch.

**Pull request merged**